docs/ROOTFS_DISCOVERY.md

```diff
@@ -0,0 +1,395 @@
+---
+title: Boot Components & Root File System Discovery
+category: Booting
+layout: default
+SPDX-License-Identifier: LGPL-2.1-or-later
+---
+
+# Boot Components & Root File System Discovery
+
+The recommended way to boot a [`systemd`](https://systemd.io/) based
+[UEFI](https://en.wikipedia.org/wiki/UEFI) system consists primarily of three
+components:
+
+1. A boot loader,
+   i.e. [`systemd-boot`](https://www.freedesktop.org/software/systemd/man/latest/systemd-boot.html)
+   that provides interactive and programmatic control of what precisely to
+   boot. It takes care of enumerating all possible boot targets, potentially
+   presenting it to the user in a menu, but otherwise picking an item
+   automatically, implementing boot counting and automatic rollback if desired.
+
+2. A unified kernel image ("UKI"), i.e. an UEFI PE executable that combines
+   [`systemd-stub`](https://www.freedesktop.org/software/systemd/man/latest/systemd-stub.html),
+   a Linux kernel, and an initial RAM disk ("`initrd`") into one. UKIs are
+   self-descriptive: the aforementioned boot loader enumerates these UKIs and
+   automatically extracts all information necessary to determine which menu
+   entries to generate for them. Within the UKI runtime the transition from the
+   UEFI firmware world into the Linux world takes place, i.e. it encapsulates
+   the fundamental `ExitBootServices()` UEFI call, that ends PC firmware
```

Comment:
mmh this happens in the kernel code though (efi stub), one might get the impression that it happens in systemd-stub when it's worded this way, so it would be better to rephrase slightly